ext/GDBM_File/GDBM_File.xs

```diff
@@ -105,31 +284,46 @@ gdbm_TIEHASH(dbtype, name, read_write, mode)
 	OUTPUT:
 	  RETVAL
 	
-
-#define gdbm_close(db)			gdbm_close(db->dbp)
-void
-gdbm_close(db)
-	GDBM_File	db
-	CLEANUP:
-
 void
 gdbm_DESTROY(db)
 	GDBM_File	db
 	PREINIT:
 	int i = store_value;
-	CODE:
-	gdbm_close(db);
+    CODE:
+        if (gdbm_file_close(db)) {
+            croak("gdbm_close: %s; %s", gdbm_strerror(gdbm_errno),
+                  strerror(errno));
+	}
 	do {
 	    if (db->filter[i])
 		SvREFCNT_dec(db->filter[i]);
 	} while (i-- > 0);
 	safefree(db);
 
+void
+gdbm_UNTIE(db, count)
+	GDBM_File	db
+        unsigned count
+    CODE:
+        if (count == 0) {
+            if (gdbm_file_close(db))
+                croak("gdbm_close: %s; %s",
+                      gdbm_strerror(gdbm_errno),
+                      strerror(errno));
+	}
+
+
```

Comment:
I don't think this is desirable.  If someone unties the hash, gdbm_file_close() is called, but it's also called by DESTROY.

Reply:
It is desirable, because it allows to catch eventual errors early. As to your second point, gdbm_file_close can be called multiple times without adverse effects: after the first call it becomes a no-op.

Reply:
ok, thanks